README.md
# nsound
nsound is a C++ library and python module for real-time sound synthesis.


ChangeLog
#------------------------------------------------------------------------------
#
# $Id: ChangeLog 926 2015-08-23 17:36:20Z weegreenblobbie $
#
#  Nsound is a C++ library and Python module for audio synthesis featuring
#  dynamic digital filters. Nsound lets you easily shape waveforms and write
#  to disk or plot them. Nsound aims to be as powerful as Csound but easy to
#  use.
#
#------------------------------------------------------------------------------

Key:
+ Finished item
- Planned item

2021-12-30 Nsound 0.9.5
    + Lots of updates for building and linking with Python3
    + Added github Actions to build on Linux, MacOs and Windows
    + Added FilterMedian
    + Updated Plotter to use latest matplotlib.pyplot
    + Added an Intersteller inspired example: src/examples/interstellar.cc


2015-08-23 Nsound 0.9.4

    + C++11 now required for compiling Nsound
    + Fixed an AudioStream pickling bug, wasn't loading n channels or sample rate
    + Switched to setuptools to build wheel packages on windows (maybe os x)
    + Moved circular_iterator out of Buffer.h
    + Fixed #76: DelayLine: add dynamic delay
    + Fixed #63: Add 'classic' filters with resonance control
        FilterLowPassMoogVcf added
    + AudioStream now using std::vector<Buffer *> instead of raw array
    + Fixed Python expressions for returning references
    + Fixed Python expressions for iterating over Buffer in AudioStreams
    + Reworked circular_iterator, added unit tests
    + Created delayline demo video: https://www.youtube.com/watch?v=9m3KA6JBDH8
    + Created tapedeck demo video: https://www.youtube.com/watch?v=Al6aHlwyEuA


2015-06-13 Nsound 0.9.3

    + Fixed #64: Nsound is now available as a source package from PyPI!
    + Fixed #74: Can't compile c++ lib if swig is not installed
    + Fixed #75: setup.py: use shell environment CC/CXX if defined
    + Some fixes to AudioPlaybackRt


2015-01-01 Nsound 0.9.2

    + Created buildbot config running on my local network [#53]
    + Fixed building on GCC 4.6 (Ubuntu 12) [#65]
    + Fixed compiling without portaudio [#66]
    + Fixed segfault with operator[] [#70]
    + Fixed python module errors with operator<< [#71]
    + AudioStreams and Buffers can now be Pickeled in Python
    + Added a CircularBuffer class which returns the last N samples on read
    + Added src/example/bebot-disp.py which adds a real-time oscilloscope and
      spectrogram display: https://www.youtube.com/watch?v=j4uLvl68PtI
    + Added StreamOperators.h to support serializing Buffers & AudioStream
      (Pickling in Python)


2014-07-01 Nsound 0.9.1

    + Real-time audio playback via AudioPlaybackRt class
        + Requires low-latency driver like JACK Audio Connection Kit for
          smooth playback
    + Added real-time playback example called piano.py
    + Added real-time playback example called bebot.py
        + Which mimics the iPhone App called 'Bebot'
    + Compiles on Mac again (#55)
    + Closed #49: Buffer::getPow() undefined in python module
    + Closed #50: AudioPlayback: use double buffering (added AudioPlaybackRt)
    + Closed #55: nsound_config: update Mac version
    + Closed #60: Filters: add a parameter 'bool is_realtime' like the Generator
        class (added setRealtime(bool) instead)
    + Closed #61: Remove 'reset_first' and add member variable is_realtime for
        the Generator class (added setRealtime(bool) instead)

2013-12-24 Nsound 0.9.0

    + Added precompiled static libs of libao & libportaudio for Windows
    + Added Visual Studio compiler again for Windows builds, 64-bit builds
    + Can now build 64-bit Python module for Windows (no cygwin\mingw required)
    + Added Visual Studio 2010+ project files (using VS 2012 Express) in msvs/
    + Refactored SCons configuration a bit (nsound_config.py).
    + Added some SCons custom builds in site_scons/site_tools
    + Fixed an init problem with Moving Average Filter
    + Added Buffer.mul() for multiplying over a sub-region of the buffer.
    + Added getFrequencyOffset(f1, octaves, cents)
    + Added kwargs to most Plotter functions
    + Added Macros.h and C++ exceptions
    + Added Python module C++ exception wrapping
    + Added Python Buffer.toList() method, converts Buffer to Python List
    + Added Utils.h, computes octaves and cents away from reference frequency
    + Improved the Vocoder band-pass filter bank, vocoder example sounds much
      better
    + Refactored some code to assert values and pointers, throwing exception on
      error
    + Refactored lots of Plotter methods, now supports kwargs, added Axes class
    + Changed Plotter::subplot(), now supports sharex, sharey (see example5.cc)
    + Renamed AdsrEnvelope to EnvelopeAdsr and reworked completely to be more
      useful, calculates sustain time automatically.
    + Removed duplicate plotting code, Nsound Python module calls the smae
      C++ code (Pylab.cc)

2012-04-15 Nsound 0.8.2

    + Fixed bug 3341096 FFTransform windowing

    + Added Spectrogram class for calculating and plotting spectrograms

    + Added AudioPlayback::getStateString()

    + Added AudioPlayback::scanDevices() which will try to play a sample on
      all the available audio drivers detected

    + Added Nsound::use(AudioBackendtype)

    + Added Nsound::getBackends(), returns a list of strings of available
      audio backends (libao or libportaudio).

    + Updated AudioBackendLibao::setOptions(), now can select specific driver

    + Updated AudioBackendLibportaudio::setOptions(), now can select specific
      driver

    + Added AudioStream::pad(), ensures channels in the AudioStream are the
      same length

    + Added AudioStream::transpose(), performs a matrix transpose by treating
      the AudioStream like a 2D matrix

    + Added AudioStream::pow() & operator^ for powers

    + Added Buffer::pow() & operator^ for powers

    + Added Buffer::log(), log10()"

    + Added Buffer::findPeaks() to find indices of local peaks


src/Nsound/Hat.cc
//-----------------------------------------------------------------------------
//
//  $Id: Hat.cc 874 2014-09-08 02:21:29Z weegreenblobbie $
//
//  Nsound is a C++ library and Python module for audio synthesis featuring
//  dynamic digital filters. Nsound lets you easily shape waveforms and write
//  to disk or plot them. Nsound aims to be as powerful as Csound but easy to
//  use.
//
//  Copyright (c) 2009-Present Nick Hilton
//
//  weegreenblobbie2_gmail_com (replace '_' with '@' and '.')
//
//-----------------------------------------------------------------------------

//-----------------------------------------------------------------------------
//
//  This program is free software; you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation; either version 2 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU Library General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//
//-----------------------------------------------------------------------------

#include <Nsound/AudioStream.h>
#include <Nsound/Buffer.h>
#include <Nsound/Hat.h>
#include <Nsound/FilterHighPassIIR.h>
#include <Nsound/Square.h>

using namespace Nsound;
using std::cout;
using std::endl;

//-----------------------------------------------------------------------------
Hat::
Hat(const float64 & sample_rate)
    :
    Instrument(sample_rate),
    square_(NULL),
    hpf_(NULL)
{
    square_ = new Square(sample_rate_);
    hpf_ = new FilterHighPassIIR(sample_rate_, 2, 4872.0, 0.05);
}

//-----------------------------------------------------------------------------
Hat::
Hat(const Hat & rhs)
    :
    Instrument(rhs.sample_rate_),
    square_(NULL),
    hpf_(NULL)
{
    square_ = new Square(*rhs.square_);
    hpf_ = new FilterHighPassIIR(*rhs.hpf_);
}

//-----------------------------------------------------------------------------
Hat::
~Hat()
{
    delete square_;
    delete hpf_;
}

//-----------------------------------------------------------------------------
Hat &
Hat::
operator=(const Hat & rhs)
{
    delete square_;
    delete hpf_;

    sample_rate_ = rhs.sample_rate_;
    square_ = new Square(*rhs.square_);
    hpf_ = new FilterHighPassIIR(*rhs.hpf_);

    return *this;
}

AudioStream
Hat::
play()
{
    AudioStream y(sample_rate_, 1);

    float64 offset = 0.124;

    y << play(1.0, 1.0, true);

    //        duration, tune, is_open
//~    y.add(play(1.0,      1.0,  true),  0 * offset);
    y.add(play(1.0,      1.0,  true),  1 * offset);
    y.add(play(1.0,      1.0,  false), 2 * offset);
    y.add(play(1.0,      1.0,  true) , 3 * offset);

    y.add(play(1.0,      1.0,  true) , 4 * offset);
    y.add(play(1.0,      1.0,  true) , 5 * offset);
    y.add(play(1.0,      1.0,  true) , 6  * offset);
    y.add(play(1.0,      1.0,  false), 7  * offset);

    y.add(play(1.0,      1.0,  true) , 8  * offset);
    y.add(play(1.0,      1.0,  true) , 9  * offset);
    y.add(play(1.0,      1.0,  false), 10 * offset);
    y.add(play(1.0,      1.0,  true) , 11 * offset);

    y.add(play(1.0,      1.0,  true) , 12 * offset);
    y.add(play(1.0,      1.0,  true) , 13 * offset);
    y.add(play(1.0,      1.0,  true) , 14 * offset);
    y.add(play(1.0,      1.0,  false), 15 * offset);

    y.add(play(1.0,      1.0,  true) , 16 * offset);
    y.add(play(1.0,      1.0,  true) , 17 * offset);
    y.add(play(1.0,      1.0,  false), 18 * offset);
    y.add(play(1.0,      1.0,  true) , 19 * offset);

    y.add(play(1.0,      1.0,  true) , 20 * offset);
    y.add(play(1.0,      1.0,  true) , 21 * offset);
    y.add(play(1.0,      1.0,  true) , 22 * offset);
    y.add(play(1.0,      1.0,  false), 23 * offset);

    y.add(play(1.0,      0.9,  true) , 24 * offset);
    y.add(play(1.0,      0.8,  true) , 25 * offset);
    y.add(play(1.0,      0.7,  true) , 26 * offset);
    y.add(play(1.0,      0.6,  true) , 27 * offset);

    y.add(play(1.0,      0.5,  true) , 28 * offset);
    y.add(play(1.0,      0.4,  true) , 29 * offset);
    y.add(play(1.0,      0.3,  true) , 30 * offset);
    y.add(play(1.0,      0.2,  true) , 31 * offset);

    return y;
}

#include <Nsound/Plotter.h>

AudioStream
Hat::
play(
    const float64 & duration,
    const float64 & tune,
    boolean is_closed)
{
    float64 play_duration = duration;
    float64 silence_duration = 0.0;

    float64 env_offset = 0.55 * play_duration;

    if(is_closed)
    {
        play_duration *= 0.06;
        silence_duration = duration - play_duration;
        env_offset = 0.0;
    }

    float64 frequency = tune * 530.0;

    Buffer env = square_->drawLine(0.0005, 0.1, 1.0)
              << square_->drawDecay(play_duration-0.0005 - env_offset, 7.0)
              << square_->silence(silence_duration + env_offset);

    AudioStream out(sample_rate_, 1);

    out << square_->generate(play_duration, frequency);

    out += square_->generate(play_duration, frequency * 1.5038);
    out += square_->generate(play_duration, frequency * 1.6132);
    out += square_->generate(play_duration, frequency * 1.9642);
    out += square_->generate(play_duration, frequency * 2.5321);
    out += square_->generate(play_duration, frequency * 2.7547);

    out /= 2.0;

    out = hpf_->filter(out);
    out = hpf_->filter(out);

    return out * env;
}

// :mode=c++: jEdit modeline
